File: synpay-dashboard-python-api/routes/departments.py
"""
Department management API routes.

These endpoints are consumed by the frontend (Next.js) and forward
all requests to the Spring Boot Integration Core for RBAC enforcement,
business logic, and audit logging.
"""

from fastapi import APIRouter, Depends, Query, Request
from fastapi.responses import JSONResponse
from pydantic import BaseModel, Field

from dependencies import extract_token, get_client_ip
from spring_client import forward_request

router = APIRouter(prefix="/api/departments", tags=["Department Management"])


# -- Request schemas -------------------------------------------------------

class CreateDepartmentBody(BaseModel):
    department_name: str = Field(..., min_length=1, max_length=100, alias="departmentName")

    model_config = {"populate_by_name": True}


class UpdateDepartmentBody(BaseModel):
    department_name: str | None = Field(None, max_length=100, alias="departmentName")

    model_config = {"populate_by_name": True}


# -- Helpers ----------------------------------------------------------------

def _spring_response(response) -> JSONResponse:
    """Convert an httpx response into a FastAPI JSONResponse."""
    return JSONResponse(
        status_code=response.status_code,
        content=response.json(),
    )


# -- GET /api/departments ---------------------------------------------------

@router.get("")
async def list_departments(
    request: Request,
    page: int = Query(0, ge=0),
    size: int = Query(20, ge=1, le=100),
    token: str = Depends(extract_token),
):
    """
    Paginated department list.
    Forwards to: GET /internal/departments
    """
    params: dict = {"page": page, "size": size}

    response = await forward_request(
        "GET",
        "/internal/departments",
        token,
        params=params,
        forwarded_for=get_client_ip(request),
        user_agent=request.headers.get("User-Agent"),
    )
    return _spring_response(response)


# -- GET /api/departments/{department_id} -----------------------------------

@router.get("/{department_id}")
async def get_department(
    department_id: int,
    request: Request,
    token: str = Depends(extract_token),
):
    """
    Get a single department by ID.
    Forwards to: GET /internal/departments/{id}
    """
    response = await forward_request(
        "GET",
        f"/internal/departments/{department_id}",
        token,
        forwarded_for=get_client_ip(request),
        user_agent=request.headers.get("User-Agent"),
    )
    return _spring_response(response)


# -- POST /api/departments --------------------------------------------------

@router.post("", status_code=201)
async def create_department(
    body: CreateDepartmentBody,
    request: Request,
    token: str = Depends(extract_token),
):
    """
    Create a new department.
    Forwards to: POST /internal/departments
    """
    payload: dict = {
        "departmentName": body.department_name,
    }

    response = await forward_request(
        "POST",
        "/internal/departments",
        token,
        json_body=payload,
        forwarded_for=get_client_ip(request),
        user_agent=request.headers.get("User-Agent"),
    )
    return _spring_response(response)


# -- PUT /api/departments/{department_id} -----------------------------------

@router.put("/{department_id}")
async def update_department(
    department_id: int,
    body: UpdateDepartmentBody,
    request: Request,
    token: str = Depends(extract_token),
):
    """
    Update an existing department.
    Forwards to: PUT /internal/departments/{id}
    """
    payload: dict = {}
    if body.department_name is not None:
        payload["departmentName"] = body.department_name

    response = await forward_request(
        "PUT",
        f"/internal/departments/{department_id}",
        token,
        json_body=payload,
        forwarded_for=get_client_ip(request),
        user_agent=request.headers.get("User-Agent"),
    )
    return _spring_response(response)


# -- DELETE /api/departments/{department_id} --------------------------------

@router.delete("/{department_id}")
async def delete_department(
    department_id: int,
    request: Request,
    token: str = Depends(extract_token),
):
    """
    Delete a department by ID.
    Forwards to: DELETE /internal/departments/{id}
    """
    response = await forward_request(
        "DELETE",
        f"/internal/departments/{department_id}",
        token,
        forwarded_for=get_client_ip(request),
        user_agent=request.headers.get("User-Agent"),
    )
    return _spring_response(response)

File: synpay-dashboard-frontend/api/departments.ts
/**
 * Department Management API Client
 *
 * Calls the FastAPI gateway at /api/departments, which forwards
 * all requests to Spring Boot Integration Core.
 *
 * RBAC is enforced server-side by Spring Boot — this layer
 * only handles HTTP transport and type-safe response mapping.
 */

import { authHeader } from '@/lib/auth'

const API_BASE = process.env.NEXT_PUBLIC_API_URL ?? 'http://localhost:8000'

// ── API Envelope ─────────────────────────────────────────────────

export interface ApiEnvelope<T> {
  success: boolean
  message?: string
  data?: T
}

// ── Response Types (mirror Spring Boot DTOs) ─────────────────────

export interface ApiDepartmentResponse {
  departmentId: number
  departmentName: string
  createdAt: string | null
  updatedAt: string | null
}

export interface ApiDepartmentPageResponse {
  content: ApiDepartmentResponse[]
  page: number
  size: number
  totalElements: number
  totalPages: number
}

// ── Request Types ────────────────────────────────────────────────

export interface CreateDepartmentPayload {
  departmentName: string
}

export interface UpdateDepartmentPayload {
  departmentName?: string
}

// ── Query Parameters ─────────────────────────────────────────────

export interface DepartmentListParams {
  page?: number
  size?: number
}

// ── API Error ────────────────────────────────────────────────────

export class ApiError extends Error {
  status: number
  constructor(message: string, status: number) {
    super(message)
    this.name = 'ApiError'
    this.status = status
  }
}

// ── Helper ───────────────────────────────────────────────────────

async function request<T>(url: string, options: RequestInit = {}): Promise<T> {
  const response = await fetch(`${API_BASE}${url}`, {
    headers: {
      'Content-Type': 'application/json',
      ...authHeader(),
      ...options.headers,
    },
    ...options,
  })

  const body: ApiEnvelope<T> = await response.json()

  if (!response.ok || !body.success) {
    throw new ApiError(
      body.message ?? `Request failed with status ${response.status}`,
      response.status,
    )
  }

  return body.data as T
}

// ── API Functions ────────────────────────────────────────────────

/** Paginated department list. Requires department.read. */
export async function fetchDepartments(
  params: DepartmentListParams = {},
): Promise<ApiDepartmentPageResponse> {
  const searchParams = new URLSearchParams()

  if (params.page !== undefined) searchParams.set('page', String(params.page))
  if (params.size !== undefined) searchParams.set('size', String(params.size))

  const qs = searchParams.toString()
  return request<ApiDepartmentPageResponse>(`/api/departments${qs ? `?${qs}` : ''}`)
}

/** Get single department by ID. Requires department.read. */
export async function fetchDepartment(id: number): Promise<ApiDepartmentResponse> {
  return request<ApiDepartmentResponse>(`/api/departments/${id}`)
}

/** Create a new department. Requires department.write. */
export async function createDepartment(
  payload: CreateDepartmentPayload,
): Promise<ApiDepartmentResponse> {
  return request<ApiDepartmentResponse>('/api/departments', {
    method: 'POST',
    body: JSON.stringify(payload),
  })
}

/** Update an existing department. Requires department.write. */
export async function updateDepartment(
  id: number,
  payload: UpdateDepartmentPayload,
): Promise<ApiDepartmentResponse> {
  return request<ApiDepartmentResponse>(`/api/departments/${id}`, {
    method: 'PUT',
    body: JSON.stringify(payload),
  })
}

/** Delete a department by ID. Requires department.write. */
export async function deleteDepartment(id: number): Promise<void> {
  return request<void>(`/api/departments/${id}`, {
    method: 'DELETE',
  })
}
